core/src/concurrent/removable_queue.hpp

```diff
@@ -0,0 +1,544 @@
+
+  if (tail_tagged != expected_tail ||
+      tail->next.load(std::memory_order_seq_cst) != expected_next) {
+    // Tail moved.
+    return GetLast();
+  }
+
+  Segment& segment = SegmentFreeList::Acquire();
+  UASSERT(segment.next.load(std::memory_order_relaxed) ==
+          TaggedPtr<Segment>(nullptr, segment.epoch));
+  TaggedPtr segment_tagged(&segment, segment.epoch);
+
+  if (!tail->next.compare_exchange_strong(expected_next, segment_tagged,
+                                          std::memory_order_seq_cst,
+                                          std::memory_order_relaxed)) {
+    // Tail moved.
+    SegmentFreeList::Release(segment);
+    return GetLast();
+  }
+
+  // Success, now try to help by fixing tail. If we fail, it means someone has
+  // already helped us.
+  tail_.compare_exchange_strong(tail_tagged, segment_tagged,
+                                std::memory_order_seq_cst,
+                                std::memory_order_relaxed);
+  return segment_tagged;
+}
+
+template <typename T>
+auto RemovableQueue<T>::RemoveFirst(TaggedPtr<Segment> expected_head) noexcept
+    -> TaggedPtr<Segment> {
+  UASSERT(expected_head.GetDataPtr() != nullptr);
+
+  if (head_.load(std::memory_order_seq_cst) != expected_head) {
+    return GetFirst();
+  }
+
+  auto* const head = expected_head.GetDataPtr();
+  const auto next_tagged = head->next.load(std::memory_order_seq_cst);
+  if (next_tagged.GetDataPtr() == nullptr) {
+    // Signal that the queue is empty.
+    return nullptr;
+  }
+
+  // tail_ either points to:
+  // 1. expected_head -> need to help once to correct to at least 'next_tagged'
+  // 2. expected_head + N -> no need to help
+  // Trying to help just once is enough.
+  auto expected_tail = expected_head;
+  if (tail_.load(std::memory_order_seq_cst) == expected_head) {
+    tail_.compare_exchange_strong(expected_tail, next_tagged,
+                                  std::memory_order_seq_cst,
+                                  std::memory_order_relaxed);
+  }
+
+  if (!head_.compare_exchange_strong(expected_head, next_tagged,
+                                     std::memory_order_seq_cst,
+                                     std::memory_order_relaxed)) {
+    return GetFirst();
+  }
+
+  DisposeSegment(*head);
+  return next_tagged;
+}
+
+template <typename T>
+auto RemovableQueue<T>::DoPush(Node& node) -> TaggedPtr<Slot> {
+  UASSERT(node.value);
+  auto last_segment_tagged = GetLast();
+
+  while (true) {
+    auto* const last_segment = last_segment_tagged.GetDataPtr();
+    UASSERT(last_segment != nullptr);
+
+    const auto segment_epoch = last_segment_tagged.GetTag();
+    const TaggedPtr<Node> node_tagged(&node, segment_epoch);
+    const TaggedPtr<Node> null_node(kNullNode, segment_epoch);
+
+    // The paper advises randomized iteration order for better performance under
+    // contention, but we trade it for better single-threaded performance.
```

Comment:
This confuses me

My understanding is that the papers is basically "Split everything in chunks of size K, randomize iteration over chunks and hope for the best", but that "randomize and hope" part is missing - don't we just have a lock-free queue that is exactly K times slower for even single-threaded performance?

Reply:
Even though the randomization is used mostly to reduce contention, it helps (it does, right?) with this loop not being O(k^2)

As of now it seems to me that we got the worst of both worlds - contention isn't reduced, complexity is O(k^2).
I propose either 

1. generating some (or even all, with K = 8 there are about 40k of them) possible permutations at compilation and selecting the permutation decently fast somehow or
2. doing some bits manipulations here to find a candidate slot in O(1)

First way should reduce contention and might be fast enough anyway, second way should be the fastest one in single-threaded performance

Reply:
The same applies to `DoTryPopStep`